### PR TITLE
Fix BLS usage

### PR DESCRIPTION
**Pull request by paulmillr**
Closes gh-3609

### CODE REVIEW

**Comment by jochem-brouwer** [comment]
Wow great, thanks a lot @paulmillr!! I will push a few commits to add Noble tests to CI and to fix the linter.

**Comment by jochem-brouwer** [comment]
Yes, will do, actually working on this PR ATM, also flipping out our test runner to use Noble by default, not MCL.

**Comment by holgerd77** [comment]
> Yes, will do, actually working on this PR ATM, also flipping out our test runner to use Noble by default, not MCL.

Wondering a bit: do we really want to fully switch to Noble for the test runner? We are also not doing this for the other pure JS stuff for performance reasons to not slow down the mostly unrelated everyday-PR-CI runs. 🤔 

The alternative would be to separate this and maybe run along the nightly VM tests. I think I would have a somewhat clear preference for that. That would give this a good balance and still ensures that we do not re-introduce new errors here.

**Comment by jochem-brouwer** [comment]
I'm not sure here. So currently (if we use MCL by default on our test runner, but Noble by default in our EVM), then we are thus showing that if one uses the alternative (MCL) the tests pass. However we want to ensure that if people use our EVM with default settings, it should also pass the state/blockchain tests.

I would rather prefer we test our default implementation (we can add a CI job either on PR or on nightly to test the alternatives, so MCL in this case). 

I just tested, the specific state tests here run 6.484 seconds using MCL, and 7.993 using Noble (using `multiplyUnsafe` already). So this is a slowdown for the test runner of ~1.6 seconds.

**Comment by paulmillr** [comment]
Another optimization has been added.

**Comment by holgerd77** [comment]
> I'm not sure here. So currently (if we use MCL by default on our test runner, but Noble by default in our EVM), then we are thus showing that if one uses the alternative (MCL) the tests pass. However we want to ensure that if people use our EVM with default settings, it should also pass the state/blockchain tests.
> 
> I would rather prefer we test our default implementation (we can add a CI job either on PR or on nightly to test the alternatives, so MCL in this case).
> 
> I just tested, the specific state tests here run 6.484 seconds using MCL, and 7.993 using Noble (using `multiplyUnsafe` already). So this is a slowdown for the test runner of ~1.6 seconds.

Yes, argumentation is also totally valid and given the numbers you are citing it is totally out of question to even consider (running MCL as default), agree!

We should also run MCL in some setup (nightly likely). This is not *totally* optional in the sense that we also provide code for this and so it's more like an option to choose from, where the functionality is already baked in. So we should also make sure to have this tested to some minimal extend.

Ah, a bit live-updated thought though: it generally would make sense to *generalize* here, also for the other optional WASM stuff, and make one test round (normal CI e.g.) all with JS and one test round (nightly e.g.) all with WASM. *Then* I am not sure though anymore how performance is affected and if this is still realistic this way around.

Anyhow: let's keep this for now, we can discuss on team call how we generally want to proceed here!

File: packages/evm/src/precompiles/bls12_381/noble.ts [hunk 1]
```diff
@@ -305,30 +321,21 @@ export class NobleBLS implements EVMBLSInterface {
...
-      const G2 = pair[1] as ProjPointType<Fp2>
-
-      if (index === 0) {
-        GT = bls12_381.pairing(G1, G2)
-      } else {
-        GT = bls12_381.fields.Fp12.mul(GT!, bls12_381.pairing(G1, G2))
-      }
-    }
+    // @ts-ignore
+    const FP12 = bls12_381.pairingBatch(pairs, true)
```
Comment (holgerd77) [inline packages/evm/src/precompiles/bls12_381/noble.ts:337]: @paulmillr 🙏

**Review by holgerd77 — APPROVED** [review]
LGTM, thanks for all the fixes and optimizations! 🙏 🙂 ❤️

**Comment by paulmillr** [comment]
@jochem-brouwer i'm curious about time of execution now after `pairingBatch`. How can one calculate this time?

> I just tested, the specific state tests here run 6.484 seconds using MCL, and 7.993 using Noble (using multiplyUnsafe already). So this is a slowdown for the test runner of ~1.6 seconds.

**Comment by jochem-brouwer** [comment]
Yes, sure! The idea is to run the state tests for both Noble and MCL. We could also run the blockchain tests, but these give blockchain overhead which we don't want.

To do so, first download the test: https://github.com/ethereumjs/ethereumjs-monorepo/issues/3609#issue-2478495412 (download the fixtures, and extract those in `./packages/ethereum-tests`.

Now run the state tests using Noble:

`npm run test:state -- --fork=Prague --dir=../fixtures/state_tests`

You can check it uses Noble, test runner outputs first:

```
BLS library used: Noble (JavaScript)
BN254 (alt_BN128) library used: Noble (JavaScript)
```

At the end of the tests, the total time is shown, for me: `Total (including setup): 7.513s`

Now run it using MCL:

`npm run test:state -- --fork=Prague --dir=../fixtures/state_tests --bls=mcl`

Output: `Total (including setup): 6.503s`

Note, this runs all tests in those fixtures, so these might include test not calling into `pairingBatch`.

**Comment by paulmillr** [comment]
Results on my M2:

```
Noble: Total (including setup): 2.879s
MCL: Total (including setup): 2.728s
```

**Pull request merged**